scripts/gitlab/hatchet-analysis.py

```diff
@@ -22,32 +23,100 @@
 
 import hatchet as ht
 
-f1 = options.report
-f2 = options.baseline
-
-gf1 = ht.GraphFrame.from_caliperreader(f1)
-gf2 = ht.GraphFrame.from_caliperreader(f2)
-
-gf3 = gf2 - gf1
-
-# Display dataframe columns
-print(gf3.dataframe.columns)
+class GenericFrame(ht.GraphFrame):
+   def __init__(self, gf):
+      generic_dataframe = gf.dataframe.copy()
+      generic_graph = gf.graph.copy()
+      generic_exc_metrics = gf.exc_metrics
+      generic_inc_metrics = gf.inc_metrics
+      generic_default_metric = gf.default_metric  # in newer Hatchet
+      generic_dataframe.iloc[0, generic_dataframe.columns.get_loc('name')] = 'Variant'
+      ii = generic_dataframe.index[0]
+      # fr = ht.frame.Frame({'name': 'Variant', 'type' : 'region'})
+      fr = ht.graphframe.Frame({'name': 'Variant', 'type': 'region'})
+      nn = ht.graphframe.Node(fr)
+      setattr(nn, '_hatchet_nid', ii._hatchet_nid)
+      setattr(nn, '_depth', ii._depth)
+      setattr(nn, 'children', ii.children)
+      generic_dataframe.rename(index={ii: nn}, inplace=True)
+      setattr(generic_graph, 'roots', [nn])
+      super().__init__(generic_graph, generic_dataframe, generic_exc_metrics, generic_inc_metrics)
+
+def ExtractCommonSubtree(gf1: ht.GraphFrame, gf2: ht.GraphFrame, metric: str) -> (ht.GraphFrame):
+   if (gf1.graph == gf2.graph):
+      return gf1
+   else:
+      cc = gf1.deepcopy()
+      cc2 = gf2.deepcopy()
+      cc.unify(cc2)
+      # search for nodes contained in both graphs {0==both, 1==left only, 2==right only}
+      filter_func = lambda x: x["_missing_node"] == 0
+      common_subtree = cc.filter(filter_func, squash=True)
+      # print(common_subtree.dataframe.columns.tolist())
+      # tt is generator object from post order tree traversal, i.e starts down at first set of leaves
+      tt = common_subtree.graph.roots[0].traverse(order="post")
+      s2 = 0.0  # sum accumulated at depth 2
+      s1 = 0.0  # sum accumulated at depth 1
+      s0 = 0.0
+      m3 = sys.float_info.max
+      # replace subtree values with sum of kernels that have run
+      for nn in tt:
+         if nn._depth == 3:
+            if common_subtree.dataframe.loc[nn, metric] < m3:
+               m3 = common_subtree.dataframe.loc[nn, metric]
```

Comment:
I am not able extract the purpose of this.

It looks like nodes with depth > 3 are ignored.
But then, a post traversal wall go through all the level 3 nodes of a given level 2 node. During this phase, m3 with be set to the minimal value encountered. Why ?

The rest appears to be accumulating values from higher to lower levels.

Reply:
Looking at the screenshot above  the tunings for Algorithm_MEMCPY are default and library, so when we compare different subtrees the leaf nodes will have different names for the tunings for the respective variant, so the best value of the tuning set is propagated up the tree, which is the minimum time. It doesn't make sense when comparing RAJA_Seq to RAJA_CUDA where the CUDA algorithms could be run with a bunch of different block sizes (so lots of tunings) and Seq only has default. Which one to compare then .. we prefer to propagate the minimum in both trees. Otherwise the trees would be too dissimilar and Hatchet again will gives red/green arrows suggesting tree too different at subtree, and nonsense will be propagated In Caliper an algorithm.tuning is the actual algorithm getting timed. Each algorithm.tuning is really an independent algorithm with a new name which we designate as algorithm.tuning.